### PR TITLE
Port QueryEngine and LocalDocumentView

### PR DESCRIPTION
**Pull request by cherylEnkidu**
#no-changelog
Port QueryEngine and LocalDocumentView

### CODE REVIEW

**Review by wu-hui — CHANGES_REQUESTED** [review]
There are some tests that are supposed to be part of this PR: QueryEngineTestCase and its subclasses. Sorry, I should have point this out earlier.

Also, please merge with master because I merged overlay into master now.

File: Firestore/core/src/core/target.cc [hunk 1]
```diff
@@ -65,6 +66,39 @@ bool Target::IsDocumentQuery() const {
 
 // MARK: - Indexing support
 
+size_t Target::GetSegmentCount() const {
+  std::set<FieldPath> fields;
```
[the remark this answers is not in the document]
Comment (cherylEnkidu) [reply Firestore/core/src/core/target.cc:70]: Looks like we don't have a hash function implemented for `FieldPath`

For code like [HashSet\<FieldPath\> maskSet = new HashSet\<\>();](https://github.com/firebase/firebase-android-sdk/blob/2bac92d248ef0fb706e732ebeec0d0d757bd47fa/firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/Mutation.java#L106) in Android SDK. The equivalent is [std::set\<FieldPath\> mask_set;](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/src/model/mutation.cc#L185) in iOS SDK. The other places are the same case.

Do you think I need to implement a Hash function for `FieldPath`?

File: Firestore/core/src/local/query_engine.h [hunk 2]
```diff
@@ -30,44 +30,84 @@ enum class LimitType;
...
+ *
+ * There are some cases where this specific optimization is not guaranteed to
+ * produce the same results as full collection scans. In these cases, query
+ * processing falls back to full scans. These cases are:
+ *
+ * - Limit queries where a document that matched the query previously no longer
+ * matches the query.
  * - Limit queries where a document edit may cause the document to sort below
- *   another document that is in the local cache.
+ * another document that is in the local cache.
```
Comment (wu-hui) [inline Firestore/core/src/local/query_engine.h:67]: This part of comment should be updated, we no longer resort to full collection scan for the first two cases, we run the query without limit using index now.

File: Firestore/core/src/local/query_engine.h [hunk 3]
```diff
@@ -30,44 +30,84 @@ enum class LimitType;
...
+  const model::DocumentMap GetDocumentsMatchingQuery(
       const core::Query& query,
       const model::SnapshotVersion& last_limbo_free_snapshot_version,
-      const model::DocumentKeySet& remote_keys);
+      const model::DocumentKeySet& remote_keys) const;
 
  private:
+  /**
+   * Performs an indexed query that evaluates the query based on a collection's
+   * persisted index values. Returns null if an index is not available.
```
Comment (wu-hui) [inline Firestore/core/src/local/query_engine.h:90]: ```Returns `nullopt` if an index ...```

Same below

**Review by wu-hui — COMMENTED** [review]
Finished reviewing the implementation code, still have the tests to review. But here are some feedbacks.

**Review by wu-hui — COMMENTED** [review]
On Android there is also SQLiteQueryEngineTest that you would need to port.

Otherwise it is looking good!

File: Firestore/core/src/local/local_documents_view.cc [hunk 4]
```diff
@@ -118,6 +123,35 @@ model::DocumentMap LocalDocumentsView::GetDocumentsMatchingCollectionGroupQuery(
   return results;
 }
 
+const LocalWriteResult LocalDocumentsView::GetNextDocuments(
```
[the remark this answers is not in the document]
Comment (cherylEnkidu) [reply Firestore/core/src/local/local_documents_view.cc:126]: I was trying to implement all functions inside LocalDocumentView. But you are right, I will remove it in the coming commit.

File: Firestore/core/src/local/query_engine.h (hunk 2 above)
Comment (cherylEnkidu) [reply Firestore/core/src/local/query_engine.h:67]: I remove the first two cases in the coming commit. Just two questions:

1. It is appropriate to add something like: `limit query will be run without limit using index` ? Would this lead to bill confusion?

2. Do I need to update Android comments as well?

File: Firestore/core/test/unit/local/leveldb_query_engine_test.cc [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+                         QueryEngineTest,
+                         testing::Values(PersistenceFactory));
+
+class LevelDbQueryEngineTest : public QueryEngineTestBase {
+ public:
+  LevelDbQueryEngineTest() : QueryEngineTestBase(PersistenceFactory()) {
+  }
+};
+
+TEST_F(LevelDbQueryEngineTest, TestCombinesIndexedWithNonIndexedResults) {
```
Comment (wu-hui) [inline Firestore/core/test/unit/local/leveldb_query_engine_test.cc:66]: We typically do not use prefix `Test` in C++ tests, just `Combines...`

**Review by wu-hui — COMMENTED** [review]
Almost there! Please check if references work as I suggested.

File: Firestore/core/src/local/leveldb_index_manager.cc [hunk 6]
```diff
@@ -455,6 +456,63 @@ absl::optional<model::FieldIndex> LevelDbIndexManager::GetFieldIndex(
...
+}
+
+const model::IndexOffset LevelDbIndexManager::GetMinOffset(
+    const std::vector<model::FieldIndex>& indexes) const {
+  HARD_ASSERT(
+      !indexes.empty(),
+      "Found empty index group when looking for least recent index offset.");
+
+  auto it = indexes.cbegin();
+  model::IndexOffset min_offset = (it++)->index_state().index_offset();
```
[the remark this answers is not in the document]
Comment (cherylEnkidu) [reply Firestore/core/src/local/leveldb_index_manager.cc:485]: The return type of `index_offset()` is const references and it doesn't make sense to change `index_offset()` behaviour. `min_offset` and `new_offset` can only capture by value or const reference, if they are const references, the reassign `min_offset = new_offset` doesn’t work :(

[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/local/leveldb_index_manager.cc:485]: I guess we can take a pointer if we really want to? 

Anyways, let's not stress on this..this is unlikely a major performance bottleneck anyways.

**Review by wu-hui — APPROVED** [review]
Thanks for the work!

**Pull request merged**